File: backend/schemas/analytics.py
"""Pydantic v2 schemas for analytics endpoints."""

from typing import Optional, List, Dict, Any
from pydantic import BaseModel


class FinancialOverview(BaseModel):
    """Financial exposure and reconciliation integrity response."""
    total_billed: str
    total_paid: str
    total_contractual_adjustments: str
    total_patient_responsibility: str
    total_variance: str
    unreconciled_amount: str
    overpayment_exposure: str
    underpayment_exposure: str
    total_denied_amount: str
    reconciliation_rate: float
    payment_rate: float
    financial_integrity_rate: float


class ClaimsKPISchema(BaseModel):
    """Claims operational KPI metrics."""
    total_claims: int
    status_distribution: Dict[str, int]
    adjudicated_claims: int
    adjudication_rate: float
    reconciled_claims: int


class PaymentKPISchema(BaseModel):
    """Payment volume and velocity KPIs."""
    total_payments_count: int
    total_paid_amount: str
    average_payment_amount: str
    zero_payment_count: int
    average_payment_turnaround_days: Optional[float] = None


class DenialKPISchema(BaseModel):
    """Denial rates and top reasons."""
    total_denials: int
    denial_rate: float
    appealable_rate: float
    top_denial_reasons: List[Dict[str, Any]]
    denial_financial_exposure: str


class QAKPISchema(BaseModel):
    """QA defect density and quality metrics."""
    total_issues: int
    issues_by_severity: Dict[str, int]
    issues_by_dimension: Dict[str, int]
    average_dq_score: float
    clean_record_rate: float
    defect_density: float


class KPIResponse(BaseModel):
    """Aggregated operational KPI container."""
    claims: ClaimsKPISchema
    payments: PaymentKPISchema
    denials: DenialKPISchema
    qa: QAKPISchema


class TrendPointSchema(BaseModel):
    """Single time-series data point."""
    time_bucket: str
    overall_dq_score: float
    dimension_scores: Dict[str, float]
    issue_count: int
    claim_volume: int


class TrendResponse(BaseModel):
    """DQ score time-series trend analysis."""
    interval: str
    points: List[TrendPointSchema]
    rolling_average_score: float
    score_velocity: float
    trend_direction: str


class RootCauseItemSchema(BaseModel):
    """Individual Pareto root cause item."""
    anomaly_category: str
    anomaly_code: str
    rule_code: str
    description: str
    severity_code: str
    dimension_code: str
    issue_count: int
    percentage_of_total: float
    cumulative_percentage: float
    financial_exposure: str


class RootCauseResponse(BaseModel):
    """Pareto 80/20 defect concentration analysis."""
    items: List[RootCauseItemSchema]
    pareto_cutoff_index: int
    primary_defect_driver: str
    total_issues_analyzed: int


class RecurrencePatternSchema(BaseModel):
    """Repeat defect pattern cluster."""
    entity_type: str
    entity_identifier: str
    anomaly_code: str
    occurrence_count: int
    first_detected_at: Optional[str] = None
    last_detected_at: Optional[str] = None
    recurrence_rank: int


class RecurrenceResponse(BaseModel):
    """Recurrence pattern clustering summary."""
    recurring_cluster_count: int
    top_repeat_entities: List[RecurrencePatternSchema]
    repeat_issue_rate: float
    total_repeating_occurrences: int


class ProviderScorecardResponse(BaseModel):
    """Provider quality and financial performance scorecard."""
    provider_id: int
    provider_reference: str
    provider_name: str
    specialty: str
    facility_id: Optional[int] = None
    facility_name: Optional[str] = None
    claim_volume: int
    total_billed: str
    total_paid: str
    payment_rate: float
    denial_rate: float
    issue_count: int
    issue_density: float
    dq_score: float
    financial_exposure: str


class PayerScorecardResponse(BaseModel):
    """Payer adjudication efficiency scorecard."""
    payer_id: int
    payer_reference: str
    payer_name: str
    payer_type: str
    claim_volume: int
    total_billed: str
    total_paid: str
    denial_rate: float
    payment_rate: float
    average_adjudication_latency_days: float
    average_payment_latency_days: float
    timely_filing_compliance_rate: float
    contractual_adjustment_ratio: float
    issue_count: int


class AnalyticsOverviewResponse(BaseModel):
    """Combined analytics overview (financial + KPIs + root causes)."""
    financial: Optional[FinancialOverview] = None
    kpis: Optional[KPIResponse] = None
    root_cause: Optional[RootCauseResponse] = None

File: frontend/src/types/analytics.ts
export interface FinancialOverview {
  total_billed: string;
  total_paid: string;
  total_contractual_adjustments: string;
  total_patient_responsibility: string;
  total_variance: string;
  unreconciled_amount: string;
  overpayment_exposure: string;
  underpayment_exposure: string;
  total_denied_amount: string;
  reconciliation_rate: number;
  payment_rate: number;
  financial_integrity_rate: number;
}

export interface ClaimsKPIs {
  total_claims: number;
  status_distribution: Record<string, number>;
  adjudicated_claims: number;
  adjudication_rate: number;
  reconciled_claims: number;
}

export interface PaymentKPIs {
  total_payments_count: number;
  total_paid_amount: string;
  average_payment_amount: string;
  zero_payment_count: number;
  average_payment_turnaround_days?: number | null;
}

export interface DenialKPIs {
  total_denials: number;
  denial_rate: number;
  appealable_rate: number;
  top_denial_reasons: Array<{
    reason_code?: string;
    count?: number;
    description?: string;
    [key: string]: unknown;
  }>;
  denial_financial_exposure: string;
}

export interface QAKPIs {
  total_issues: number;
  issues_by_severity: Record<string, number>;
  issues_by_dimension: Record<string, number>;
  average_dq_score: number;
  clean_record_rate: number;
  defect_density: number;
}

export interface KPIOverview {
  claims: ClaimsKPIs;
  payments: PaymentKPIs;
  denials: DenialKPIs;
  qa: QAKPIs;
}

export interface DQTrendPoint {
  time_bucket: string;
  overall_dq_score: number;
  dimension_scores: Record<string, number>;
  issue_count: number;
  claim_volume: number;
}

export interface DQTrendsSummary {
  interval: string;
  points: DQTrendPoint[];
  rolling_average_score: number;
  score_velocity: number;
  trend_direction: string;
}

export interface RootCauseItem {
  anomaly_category: string;
  anomaly_code: string;
  rule_code: string;
  description: string;
  severity_code: string;
  dimension_code: string;
  issue_count: number;
  percentage_of_total: number;
  cumulative_percentage: number;
  financial_exposure: string;
}

export interface RootCauseResponse {
  items: RootCauseItem[];
  pareto_cutoff_index: number;
  primary_defect_driver: string;
  total_issues_analyzed: number;
}

export interface RecurrencePattern {
  entity_type: string;
  entity_identifier: string;
  anomaly_code: string;
  occurrence_count: number;
  first_detected_at?: string | null;
  last_detected_at?: string | null;
  recurrence_rank: number;
}

export interface RecurrenceResponse {
  recurring_cluster_count: number;
  top_repeat_entities: RecurrencePattern[];
  repeat_issue_rate: number;
  total_repeating_occurrences: number;
}

export interface ProviderScorecard {
  provider_id: number;
  provider_reference: string;
  provider_name: string;
  specialty: string;
  facility_id?: number | null;
  facility_name?: string | null;
  claim_volume: number;
  total_billed: string;
  total_paid: string;
  payment_rate: number;
  denial_rate: number;
  issue_count: number;
  issue_density: number;
  dq_score: number;
  financial_exposure: string;
}

export interface PayerScorecard {
  payer_id: number;
  payer_reference: string;
  payer_name: string;
  payer_type: string;
  claim_volume: number;
  total_billed: string;
  total_paid: string;
  denial_rate: number;
  payment_rate: number;
  average_adjudication_latency_days: number;
  average_payment_latency_days: number;
  timely_filing_compliance_rate: number;
  contractual_adjustment_ratio: number;
  issue_count: number;
}

export interface AnalyticsOverview {
  financial?: FinancialOverview | null;
  kpis?: KPIOverview | null;
  root_cause?: RootCauseResponse | null;
}
